solutions/usaco-1206.mdx

```diff
@@ -0,0 +1,120 @@
+---
+id: usaco-1206
+source: USACO Silver 2022 February
+title: Redistributing Gifts
+author: Chongtian Ma
+---
+
+[Official Analysis](http://www.usaco.org/current/data/sol_prob1_silver_feb22.html)
+
+## Explanation
+
+Similar to the official analysis, we can construct a directed edge from $i$ to $j$ if cow $i$ prefers gift $j$ to gift $i$.
+If cow $i$ is able to receive a gift where it has an outgoing edge, then cow $i$ has received the most optimal gift it can be reassigned.
+Remember to also add an edge from cow $i$ to itself in case it cannot get a better gift.
+
+Then, we can separate the graph into strongly connected components (SCCs). If $i$ is in an SCC, this means that cow $i$ can receive
```

Comment:
what's the intuition behind this?

Reply:
It’s just implementation beauty since running kosaraju you are guaranteed to least one node with an outgoing edge (itself). Should i mention it? I didn’t becuase i thought it was kind of redundant.

Reply:
Also it helps in the case where it has the original gift already at the most preferred gift.